File: others/3000-3100/3072. 将元素分配到两个数组中 II.cpp
//
// Created by 123 on 25-7-21.
//
#include <algorithm>

#include "bits/stdc++.h"
using namespace std;
using ll = long long;


class FenwickTree {
public:
	vector<int> tree;

	FenwickTree(int N) {
		tree.resize(N + 1);
	}

	void update(int index) {
		for (int i = index + 1; i < tree.size(); i += i & -i) {
			tree[i]++;
		}
	}

	int preSum(int index) {
		int s = 0;
		for (int i = index; i > 0; i -= i & -i) {
			s += tree[i];
		}
		return s;
	}
};

class Solution {
public:
	vector<int> resultArray(vector<int> &nums) {
		int n = nums.size();
		auto sorted = nums;
		ranges::sort(sorted);
		sorted.erase(unique(sorted.begin(), sorted.end()), sorted.end());
		int m = sorted.size();

		vector a{nums[0]};
		vector b{nums[1]};

		FenwickTree ft1{m};
		FenwickTree ft2{m};
		auto it1 = ranges::lower_bound(sorted, nums[0]) - sorted.begin();
		auto it2 = ranges::lower_bound(sorted, nums[1]) - sorted.begin();
		ft1.update(it1);
		ft2.update(it2);

		for (int i = 2; i < n; i++) {
			int x = nums[i];
			int t = ranges::lower_bound(sorted, x) - sorted.begin();
			int s1 = a.size() - ft1.preSum(t + 1);
			int s2 = b.size() - ft2.preSum(t + 1);

			if (s1 < s2 or (s1 == s2 and b.size() < a.size())) {
				ft2.update(t);
				b.emplace_back(x);
			} else {
				ft1.update(t);
				a.emplace_back(x);
			}
		}
		a.insert(a.end(), b.begin(), b.end());
		return a;
	}
};

File: others/3000-3100/3072. 将元素分配到两个数组中 II.py
from bisect import bisect_left
from typing import List


class FenwickTree:
    def __init__(self, n: int) -> None:
        self.tree = [0] * (n + 1)

    def update(self, index: int, val: int) -> None:
        i = index + 1
        while i < len(self.tree):
            self.tree[i] += 1
            i += i & -i

    def pre(self, index: int) -> int:
        s = 0
        while index > 0:
            s += self.tree[index]
            index -= index & -index
        return s

    def query(self, index: int) -> int:
        return self.pre(index + 1)


class Solution:
    def resultArray(self, nums: List[int]) -> List[int]:
        arr1 = [nums[0]]
        arr2 = [nums[1]]

        temp = sorted(list(set(nums)))
        m = len(temp)
        n = len(nums)

        ft1 = FenwickTree(m)
        ft2 = FenwickTree(m)

        p1 = bisect_left(temp, nums[0])
        p2 = bisect_left(temp, nums[1])

        ft1.update(p1, 1)
        ft2.update(p2, 1)

        for i in range(2, n):
            x = nums[i]
            p = bisect_left(temp, x)

            d1 = len(arr1) - ft1.query(p)
            d2 = len(arr2) - ft2.query(p)

            if d1 < d2 or (d1 == d2 and len(arr2) < len(arr1)):
                arr2.append(x)
                ft2.update(p, 1)
            else:
                arr1.append(x)
                ft1.update(p, 1)
        return arr1 + arr2

